### ISSUE TITLE
scMerge 

### ISSUE THREAD

**Issue by kevinwang09** [Bioconductor/Contributions#1014]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/SydneyBioX/scMerge

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by kevinwang09** [comment]
Hi, the only warning at this point is that we exceeded 5 minutes when running R CMD check… (Somehow Windows and Mac took much longer to run than Linux). We are also not sure why "BUILD BIN" was skipped on Linux. Any help would be appreciated. Thanks!

**Comment by nturaga** [comment]
# scMerge Review

## DESCRIPTION

* No need to use `Maintainer` and `Authors@R`. Just Authors@R is
  enough.

* Description needs to be like a paper abstract. Enough information
  needs to be provided

## NAMESPACE

* It's easier to import specific functions in the NAMESPACE of large
  packages like BiocParallel.

## man

* Some of the man pages need more description. The major functions
  have good man pages.

## Vignettes

* Include a package installation step as
  well. `BiocManager::install("scMerge")`.

* The title of the vignette doesn't load. Please check your
  vignette.

* What are the other Rmd files in the vignettes folder,

```
 Breast_Data
 CellBench_Data
 ESC_Data
 Mouse_Liver_Data
 Neuronal_Data
 Pancreas4_Data
 mESC_Data
```

These are not linked to the vignettes, because when the user does
`browseVignettes('scMerge') these sources don't show up. It's likely
because they need to be in the vignettes parent folder and not sub
folders. Please check this.

* The CSS styling gets displayed when the vignette is run using,

		R CMD Stangle scMerge.Rmd
		
		R -e "source('vignettes/scMerge.R', max=Inf, echo=TRUE)"
 
 .....
 
		> source("vignettes/scMerge.R", max=Inf, echo=TRUE)

		> ## ----style, echo = FALSE, results = 'asis'-------------------------------
		> BiocStyle::markdown()
		<script type="text/javascript">
		document.addEventListener("DOMContentLoaded", function() {
		document.querySelector("h1").className = "title";
		});
		</script>
		<script type="text/javascript">
		document.addEventListener("DOMContentLoaded", function() {
		var links = document.links;
		for (var i = 0, linksLength = links.length; i < linksLength; i++)
			if (links[i].hostname != window.location.hostname)
				links[i].target = '_blank';
					});
					</script>
					<style type="text/css" scoped>
					body {
		margin: 0px auto;
		max-width: 1134px;
		font-family: sans-serif;
		font-size: 10pt;
		}


## inst

* Please make sure the NEWS files is parsable by `utils::news()`

## tests

* the tests you have for your package are not sufficient. Consider
  writing test cases for the R code, per function. Tests need to be
  robust, and have code coverage. 
  
  One good way to write test cases is break up your functions into
  smaller testable units.
  
  eg:
  
		main <- function(a, b) {
			res <- add_two(a,b)
			message("result of addition: ", res)
			res1 <- complex_integrated_function(res, a, b)
			res * res1
		}

		add_two <- function(a, b) { a + b }

	test case:
	
		test_add_two <- function() {
			a <- 5
			b <- 5
			expect_equal(a + b, 10)
		}
		
	Do this through out your code base. Break them up into testable
    small units.
	
## docs

* Not sure why you have this folder as part of your R-package. 


## R

* Don't use "print" as an end user message in your code.

		End-User messages

		message() communicates diagnostic messages (e.g., progress during lengthy computations) during code evaluation.
		warning() communicates unusual situations handled by your code.
		stop() indicates an error condition.
		cat() or print() are used only when displaying an object to the user, e.g., in a show method.

http://bioconductor.org/developers/how-to/coding-style/

* You don't need a loop for this,

		cell_type <- apply(M, 1, function(x) which(x == 1))

	eg:

		which(diag(11) == 1)
		
* You need to better indent your code. Please follow formatting used
  in formatR.
  
* Use `<-` when defining functions. 

		calculateSil = function(x, fast_svd, cell_type, batch)
		
* Remove commented out code.

**Comment by nturaga** [comment]
When you are ready for a next review, please leave a detailed message about the changes you made.

**Comment by kevinwang09** [comment]
@nturaga 

Thank you for your review comments. We have updated the `scMerge` package to Version 0.99.22. A summary of the updates: 

1. The package description, news, namespace, installation instruction and the manuals were updated based on your comments. We hope this is sufficient to give the end-users a concise summary of the package and each implemented functions.

1. With regard to the vignette. We have fixed the title issue and it should now load with `vignette(“scMerge”)`.

1. Files like `Breast_Data.Rmd` are included purely for the construction of our `pkgdown`-GitHub webpage (https://sydneybiox.github.io/scMerge/). These files should be ignored. The only vignette file we intend for our package is `vignette/scMerge.Rmd`. The `pkgdown` package automatically creates a “docs” folder, which, to the best of our understanding, does not install to user’s computer. 

1. With regards to the print methods, we have implemented your suggestion. We have consolidated all diagnostic print outs under a verbose option. 

1. With regards to testing scripts, thank you for this comment. We have included these scripts now and the code coverage at this point is 85%. 

1. With regards to computational speed-ups, we have done another check of our codes to make sure vectorisation is used wherever possible. 

1. With regards to code styling, we have used the `formatR` package with some manual editing at places where it fails to correct for line-character length. 

Thank you, 
Best wishes, 
Kevin Wang on behalf of all authors

**Comment by nturaga** [comment]
Hi Kevin,

Some updates, and things to consider,

Overall your package looks good. It should be accepted for the RELEASE on
the 24th.

1. I left a review item, mentioning that `<-` is more appropriate as far as
R goes when defining top level functions. Please consider making this
change.

2. Looking at your `git log` it's very hard to determine what changes
have been made specifically when you say "vectorization". Because your
commits have been mixed with formatting changes and possible optimizations
in the code. I've had to go through your codebase all over again. For the
next time, it's useful to specifically reply to each point in the review
showing the code changes or linking to them.

Best,

Nitesh

**Comment by kevinwang09** [comment]
Hi Nitesh, 
Thank you for your feedback. I have now corrected the assignment issue in V0.99.23. And I will take care of the commit message in the future. The reason for the unclear messages is because I was working on the review at the same time as some code-simplification.

Thank you
Kevin

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/kevinwang09.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("scMerge")`. The package 'landing page' will be created at

https://bioconductor.org/packages/scMerge

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/

**Comment by kevinwang09** [comment]
Hi @mtmorgan / @nturaga 
I just added my ssh key to Bioconductor, and 

+ I can see https://github.com/kevinwang09.keys is not empty. 
+ I already ran `git remote add upstream git@git.bioconductor.org:packages/scMerge.git` in my local git repo. 
+ `git remote -v` shows expected outputs:

`origin	https://github.com/SydneyBioX/scMerge (fetch)`
`origin	https://github.com/SydneyBioX/scMerge (push)`
`upstream	git@git.bioconductor.org:packages/scMerge.git (fetch)`
`upstream	git@git.bioconductor.org:packages/scMerge.git (push)`

+ Running `ssh -T git@git.bioconductor.org` also shows the expected output of `R W  	packages/scMerge`

+ However, running `git fetch upstream` and `git remote show upstream` shows an error:

`fatal: protocol error: bad line length character: Usin`. 

And I cannot resolve this problem looking at this thread https://stackoverflow.com/questions/8170436/git-remote-error-fatal-protocol-error-bad-line-length-character-unab. My `~/.bashrc` only has a single line `export PATH=/usr/local/bin:$PATH` and I am using Mac OS. 

Any help would be appreciated. Thanks, Kevin.
